### PR TITLE
Improve E0401 diagnostics to mention other items

### PR DESCRIPTION
**Pull request by Manishearth**
r? @steveklabnik

### CODE REVIEW

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -286,12 +286,32 @@ fn foo<T>(x: T) {
 }
 ```
 
-Functions inside functions are basically just like top-level functions, except
-that they can only be called from the function they are in.
+nor will this
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: I think you forgot ':' at the end of your sentence.

File: src/librustc_resolve/diagnostics.rs [hunk 2]
```diff
@@ -286,12 +286,32 @@ fn foo<T>(x: T) {
...
+}
+```
+
+Items inside functions are basically just like top-level items, except
+that they can only be used from the function they are in.
 
 There are a couple of solutions for this.
 
-You can use a closure:
+If the item is function, you may use a closure:
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: "If the item is a function" perhaps?

File: src/librustc_resolve/diagnostics.rs [hunk 3]
```diff
@@ -313,6 +333,13 @@ fn foo<T>(x: T) {
 }
 ```
 
+```
+fn foo<T>(x: T) {
+    type MaybeT<T> = Option<T>;
+}
+```
+
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: It's a lot of empty lines!

**Comment by Manishearth** [comment]
fixed

**Comment by nagisa** [comment]
ping @steveklabnik

**Pull request merged**